### ISSUE TITLE
YOLO26N INT8 Quantization & Layer Extraction

### ISSUE THREAD

**Issue by divinemathew** [ultralytics/ultralytics#24469]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hi am planning to infer the YOLO26n in FPGA which needs a QUANTIZED INT8 Model so that the weights and bias of each layer can be extracted. 

Is there a way to do that ? Can you guys please help in doing that ..

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @divinemathew, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue routed quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training or model export ❓ Question, please provide as much information as possible, including your target workflow, export steps attempted, logs, current environment details, and any relevant model artifact details, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by raimbekovm** [comment]
@divinemathew yes, this is fully supported. For FPGA deployment you want a format that gives you direct access to per-layer INT8 weights, biases, and per-channel quantization scales/zero-points. TFLite INT8 is the standard choice.

## Export

```python
from ultralytics import YOLO

model = YOLO("yolo26n.pt")
model.export(format="tflite", int8=True, data="coco.yaml", imgsz=320)
```

This produces several variants in `yolo26n_saved_model/`. The two you care about are:

- `yolo26n_int8.tflite` - INT8 weights, float32 I/O
- `yolo26n_full_integer_quant.tflite` - INT8 weights and INT8 I/O (typically what you want on FPGA, so the host doesn't have to (de)quantize at the boundary)

## Extracting weights and biases per layer

Just use the TFLite Python interpreter (or any flatbuffer reader, since `.tflite` is a flatbuffer):

```python
import numpy as np
from tensorflow.lite.python.interpreter import Interpreter

itp = Interpreter(model_path="yolo26n_saved_model/yolo26n_full_integer_quant.tflite")
itp.allocate_tensors()

for t in itp.get_tensor_details():
    if t["dtype"] is np.int8 and len(t["shape"]) >= 2:
        w = itp.get_tensor(t["index"])              # raw int8 weights
        qp = t["quantization_parameters"]
        scales = qp["scales"]                        # per-channel scales
        zero_points = qp["zero_points"]              # per-channel zero points
        # dequantized float = (w - zero_point) * scale, applied per output channel
        print(t["name"], w.shape, w.dtype, "scales:", scales.shape)
```

I just ran this on a fresh export of `yolo26n` and got 94 INT8 weight tensors. Conv weights are stored as `(out_channels, kH, kW, in_channels)` (NHWC), and biases are INT32 with their own scales. For `full_integer_quant`, input is `int8` with `scale=0.003922, zp=-128`, so `int8 = round(float / scale + zp)` is what you feed.

## Alternatives worth a look

- **ONNX + onnxruntime.quantization** if your FPGA toolchain prefers ONNX QDQ format
- **OpenVINO INT8** (`format="openvino", int8=True`) produces `.xml` + `.bin` where weights are readable
- **IMX** (`format="imx", int8=True`) if your target is Sony's IMX500 NPU/sensor specifically

Full INT8 quantization context, including calibration data tips, is in the [TensorRT INT8 docs](https://docs.ultralytics.com/integrations/tensorrt/#using-tensorrt-int8-quantization) (the calibration guidance generalizes to TFLite too: provide a representative `data=...` dataset).

**Comment by AffanBinFaisal** [comment]
Just adding a few small notes on top of @raimbekovm's answer (TFLite is indeed the easiest path for FPGA):

1. **Heads up: `format="onnx", int8=True` does nothing in Ultralytics right now.** Only these formats actually do INT8: `openvino, engine, coreml, tflite, tfjs, mnn, imx, axelera` (see [`default.yaml`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml)). The ONNX export code in [`engine/exporter.py`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/exporter.py) just ignores the `int8` flag. So if you really need an INT8 ONNX file, export a normal FP32 ONNX first and then quantize it yourself with `onnxruntime.quantization.quantize_static(...)`.

2. `format="engine", int8=True` (TensorRT) gives you fast inference, but the resulting file is a sealed binary - you cannot read out per-layer weights from it. Not useful for FPGA.

3. Calibration data quality really matters here. Pass a `data=...` YAML that looks like your real input, and use at least ~300 images (the exporter prints a warning if you give it fewer).

Which FPGA tool are you using - Xilinx Vitis AI, Intel OpenVINO-on-FPGA, or a custom HLS flow? That decides whether TFLite full-integer, OpenVINO IR, or your-own-quantized ONNX is the best input format.

**Comment by divinemathew** [comment]
@AffanBinFaisal  @raimbekovm 

Thanks for your valuable time and responding to me.


From the above comments I was able to understand that **ONNX model is not fully able to quantise** and try to use **tflite.**

@AffanBinFaisal  - I am planning to use Xilinx Vivado for FPGA implementation. 

@raimbekovm - My plan is to implement the entire YOLO26n model (Back-Bone, Neck, Head) in the FPGA. For this my initial huddle is to extract the layer structure and tensors in INT8. By using **tflite** is it possible to do the same ?

Also is it possible to extract layers of the **tflite** model and implement that in FPGA using DSP's ?

I am attaching a sample reference extraction of layers (yolo26_layerwise_math.html) which I have done using ONNX FP32. 

[yolo26_layerwise_math.html](https://github.com/user-attachments/files/27806323/yolo26_layerwise_math.html)

**Comment by divinemathew** [comment]
@AffanBinFaisal 

I am planning to implement the entire YOLO26n framework inside the FPGA using FPGA resources. So can you please point out to me which will be the good approach

**Comment by AffanBinFaisal** [comment]
@divinemathew since you are targeting Xilinx/AMD FPGA resources, the best approach depends strongly on the exact FPGA/board.

If your target device supports the AMD Vitis AI flow, I would recommend starting there rather than manually rebuilding the whole YOLO26n network in Vivado RTL/HLS. Vitis AI is designed for this kind of deployment: it supports BYOM workflows for PyTorch/TensorFlow/ONNX-like models, handles graph management, quantization and compilation, and maps supported subgraphs to AMD NPU/DPU-style accelerator flows depending on the target.

References:
- AMD Vitis AI BYOM: https://vitisai.docs.amd.com/en/latest/docs/customization_opportunities/byom.html
- AMD Vitis AI overview/compiler flow: https://vitisai.docs.amd.com/en/latest/

For manual Vivado/Vitis HLS implementation, TFLite INT8 is still useful as an extraction/reference format. A full-integer TFLite export can expose layer tensors, shapes, INT8 weights, INT32 biases, scales and zero-points through the TFLite interpreter/flatbuffer. But TFLite will not directly generate Vivado hardware. You would still need to implement the hardware architecture yourself: INT8 convolution/MAC datapaths using DSPs, feature-map buffering, layer scheduling, concat/upsample blocks, detect head logic, quantization scale handling, memory bandwidth, and post-processing/NMS if you want that inside FPGA too.

References:
- Ultralytics TFLite export arguments, including `int8`, `data`, and `fraction`: https://docs.ultralytics.com/integrations/tflite/#export-arguments
- TensorFlow Lite Interpreter tensor details and `quantization_parameters`: https://www.tensorflow.org/api_docs/python/tf/lite/Interpreter

So I would suggest:

1. First identify the exact Xilinx board/device.
2. If Vitis AI/NPU/DPU support exists for that target, use that route first.
3. If this is a fully custom accelerator/research implementation, use TFLite full-integer INT8 mainly as a reference/extraction format, then validate the FPGA implementation layer-by-layer against TFLite outputs.

Also note that `format="onnx", int8=True` is not currently the direct Ultralytics INT8 ONNX path. The Ultralytics export table lists `int8` for formats such as OpenVINO, TensorRT, SavedModel, TFLite, TF.js, MNN, IMX, Axelera and DeepX, but not for ONNX. If an ONNX-based Xilinx flow is required, export FP32 ONNX first and then use the AMD/Vitis quantization flow or another ONNX quantization tool supported by your target.

Reference:
- Ultralytics supported export formats and arguments: https://docs.ultralytics.com/integrations/#export-formats


@raimbekovm Please correct me if I am wrong.

**Comment by divinemathew** [comment]
@AffanBinFaisal 

Thank you for the support. I will look into it and come back 

Thanks for the support

**Comment by onuralpszr** [comment]
I am closing the issue for now, feel free to re-open if you guys needed, thank you for help.

**Comment by divinemathew** [comment]
@AffanBinFaisal 

I am trying to target AMD KRIA. 

Can you guys please guide to quantise and compile the Yolo26 for this
